assembly/runtime.ts

```diff
@@ -845,125 +894,153 @@ export class RootContext extends BaseContext {
+  }
 
   onRequestHeaders(a: u32): FilterHeadersStatusValues { return FilterHeadersStatusValues.Continue }
   onRequestMetadata(a: u32): FilterMetadataStatusValues { return FilterMetadataStatusValues.Continue }
-  onRequestBody(body_buffer_length: size_t, end_of_stream: bool): FilterDataStatusValues { return FilterDataStatusValues.Continue }
+  onRequestBody(body_buffer_length: usize, end_of_stream: bool): FilterDataStatusValues { return FilterDataStatusValues.Continue }
   onRequestTrailers(a: u32): FilterTrailersStatusValues { return FilterTrailersStatusValues.Continue }
   onResponseHeaders(a: u32): FilterHeadersStatusValues { return FilterHeadersStatusValues.Continue }
   onResponseMetadata(a: u32): FilterMetadataStatusValues { return FilterMetadataStatusValues.Continue }
-  onResponseBody(body_buffer_length: size_t, end_of_stream: bool): FilterDataStatusValues { return FilterDataStatusValues.Continue }
+  onResponseBody(body_buffer_length: usize, end_of_stream: bool): FilterDataStatusValues { return FilterDataStatusValues.Continue }
   onResponseTrailers(s: u32): FilterTrailersStatusValues { return FilterTrailersStatusValues.Continue }
-  onLog(): void { }  // Called after onDone when logging is requested.
+  
+  // Called after onDone when logging is requested.
+  onLog(): void { 
+    log(LogLevelValues.debug, "context id: " + this.context_id.toString() + ": onLog()");
+  }
 }
 
 function get_plugin_root_id(): string {
-
   let root_id = get_property("plugin_root_id");
   if (root_id.byteLength == 0) {
     return "";
   }
   return String.UTF8.decode(root_id);
 }
 
+let root_context_factory_map = new Map<string, (context_id: u32) => RootContext>();
+
 let context_map = new Map<u32, BaseContext>();
+
+//create root context if doesn't exist
 export function ensureRootContext(root_context_id: u32): RootContext {
+  log(LogLevelValues.debug, "ensureRootContext(root_context_id: " + root_context_id.toString() + ")");
+  log(LogLevelValues.debug, "Current context_map: " + context_map.keys().join(", "));
   if (context_map.has(root_context_id)) {
+    log(LogLevelValues.debug, "Returning root context for id: " + root_context_id.toString());
     return getRootContext(root_context_id);
   }
-  let root_id = get_plugin_root_id();
-  if (root_factory.has(root_id)) {
-    let root_context_func = root_factory.get(root_id);
-    let root_context = root_context_func();
-    root_context.context_id = root_context_id;
-    context_map.set(root_context_id, root_context);
-
-    log(LogLevelValues.warn, "returning context for " + root_id);
-    return root_context;
+  const root_id = get_plugin_root_id();
+  log(LogLevelValues.debug, "Registering new root context for " + root_id + " with id: " + root_context_id.toString());
+  if (!root_context_factory_map.has(root_id)) {
+    throw new Error("Missing root context factory for root id: " + root_id);
```

Comment:
note that AS doesnt really support exceptions now and they trigger a panic.
i think it is ok to panic here, as this should never happen, so no objections from me. wanted to make sure it works for you too.

Reply:
Yeah, I know. It is such a pain not being able to use exceptions. I hope they add such features soon because it really limits the debugging.

Reply:
BTW, I only added thow declarations when the Filter implementation is incorrect.